### PR TITLE
Support Ruby 2.3

### PR DESCRIPTION
**Pull request by mathieujobin**
my work in progress attempt to support both ruby 2.3 and 2.4
solving issue https://github.com/oggy/looksee/issues/35

I am about to run out of time tonight... so I thought of sharing my progress

### CODE REVIEW

**Comment by mathieujobin** [comment]
tired... but this is where I am at

image

**Comment by mathieujobin** [comment]
i don't know why the casts to `st_data_t` on the calls to `st_foreach` dont work in 2.3.x and 2.4.x
strangely,... I could remove the cast for other rubies and everything kept passing ?!?

image

**Comment by mathieujobin** [comment]
my build on my fork

image

the PR build

image

another puzzle

**Comment by mathieujobin** [comment]
need help here !

File: ext/mri/2.3.0/id_table.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#ifndef RUBY_ID_TABLE_H
+#define RUBY_ID_TABLE_H 1
+#include "ruby/ruby.h"
+
+struct rb_id_table;
```
Comment (manveru) [inline ext/mri/2.3.0/id_table.h]: maybe you need `typedef struct rb_id_table;`?

Comment (manveru) [inline ext/mri/2.3.0/id_table.h]: see also http://stackoverflow.com/questions/2700646/dereferencing-pointer-to-incomplete-type

Comment (mathieujobin) [inline ext/mri/2.3.0/id_table.h]: I tried casting using this struct instead, but it really expect a `st_data_t`.

```
compiling mri/mri.c
mri/mri.c: In function ‘internal_instance_methods’:
mri/mri.c:104:51: error: dereferencing pointer to incomplete type
mri/mri.c:104:89: warning: passing argument 3 of ‘st_foreach’ makes integer from pointer without a cast [enabled by default]
/home/travis/.rvm/rubies/ruby-2.3.1/include/ruby-2.3.0/ruby/st.h:122:5: note: expected ‘st_data_t’ but argument is of type ‘struct rb_id_table *’
mri/mri.c: In function ‘Looksee_internal_undefined_instance_methods’:
mri/mri.c:142:51: error: dereferencing pointer to incomplete type
mri/mri.c:142:90: warning: passing argument 3 of ‘st_foreach’ makes integer from pointer without a cast [enabled by default]
/home/travis/.rvm/rubies/ruby-2.3.1/include/ruby-2.3.0/ruby/st.h:122:5: note: expected ‘st_data_t’ but argument is of type ‘struct rb_id_table *’
make: *** [mri.o] Error 1
```

https://travis-ci.org/mathieujobin/looksee/jobs/138376615

if a typedef is missing in this file, would that be a bug in ruby 2.3 ? 

I don't know what this empty struct does? I don't remember seeing struct with no members.

Thanks for taking a look at this, much appreciated.

Comment (manveru) [inline ext/mri/2.3.0/id_table.h]: Yeah, I'm sorry that I won't be much help here, since I don't write C much either.

**Comment by josephks** [comment]
Until this fix is in can the gemspec for the current version be modified so bundler doesn't attempt to install looksee on Rubies > 2.2 ?

**Comment by mathieujobin** [comment]
@josephks https://github.com/oggy/looksee/pull/38

**Comment by mathieujobin** [comment]
still a work in progress, but I understand a lot more what is going on....

**Pull request closed without merging**